sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -441,6 +537,17 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  /**
+   * (Java-specific)
+   * Collects the elements to a Java list.
+   *
+   * Due to the incompatibility problem between Scala and Java, the return type of [[collect()]] at
```

Comment:
This just means that the RDD has the wrong classtag.  We need to find a way to pass the classtag from the encoder before calling collect.